### ISSUE TITLE
Migrating PeptideMass, PeptideDecoder, and PeptideEncoder from depthcharge v0.2.3 to casanovo

### ISSUE THREAD

**Issue by CCranney** [Noble-Lab/casanovo#337]
Hi all,

I noticed that casanovo cannot utilize versions of depthcharge above v0.2.3, largely because several key classes were removed between v0.2.3 and v0.3 of depthcharge (specifically the PeptideMass, PeptideDecoder, and PeptideEncoder classes). I assume there has been some exchange on the separation between casanovo and depthcharge, but have not been able to find it in the github discussions or issues pages. **_If that discussion exists and this is a redundant request, please let me know._**

Correct me if I'm wrong, but possible future developments - such as #321 - would require casanovo to utilize higher versions of depthcharge, so this discrepancy would need to be addressed eventually. I see one possible solution being the migration of those lost classes from depthcharge v0.2.3 to casanovo so they can continue to be maintained independently of depthcharge. I'm making this issue to ask if that would be a desirable outcome, as well as to track the necessary updates leading up to a pull request if it is.

I tinkered with migrating the requisite files while maintaining their respective commit histories, and managed to do so [here](https://github.com/CCranney/casanovo/tree/dev/casanovo/denovo) (specifically files `peptide_transformers.py` and `masses.py`). The location and file names were chosen arbitrarily as I experimented with the migration process, but I'd imagine you would want them in specific locations.

If this is a feature you would like to implement, I can make the requisite changes (including migrating the unit tests) and create a pull request. Before doing that I would primarily need to know how you would like them saved, as changing the location of the files in a pull request while maintaining commit history is difficult. Perhaps something like the following?

```
.
├── data
└── denovo
└── peptides
    └── masses.py
    └── transformers.py

```

Let me know,
Caleb

**Comment by bittremieux** [comment]
You're right that Casanovo is currently limited to DepthCharge v0.2.x. This is currently by design.

DepthCharge has undergone several breaking changes in v0.3 and v0.4, which will require significant refactoring of Casanovo. However, these changes are all based on making DepthCharge more powerful and more flexible. So there's not really functionality lost in DepthCharge, if anything, there's more functionality now (i.e. support for small molecules).

One of our major tasks in the next few weeks/months will be to upgrade the DepthCharge version used in Casanovo and then build on this new functionality to [improve Casanovo](https://github.com/Noble-Lab/casanovo/milestone/1). This will take significant effort under the hood, but should be transparent for the user at first, then resulting in improved performance later.

**Comment by bittremieux** [comment]
It's far from an easy fix, there will need to be significant refactoring. But because it deals with the internals of Casanovo that users are typically not exposed to, we hadn't created an explicit issue yet. Instead, we were tracking progress on the DepthCharge updates in our regular team meetings, although this is indeed a bit opaque for external developers.

**Comment by bittremieux** [comment]
Fixed in #442.
